### ISSUE TITLE
fetch timed out

### ISSUE THREAD

**Issue by esovetkin** [tonarino/innernet#89]
I have a weird issue, that started soon after I setup several nodes in my innernet.

I have a setup with 8 nodes, 2 of which are behind NAT1, 5 are behind NAT2, and 1 server node that has a public static ip address. I set up my innernet-server, then joined an admin node (`gellert`) that is behind NAT1, from which I setup all other nodes. I have two cidr subnetworks, and issue have started after add-association between from two cidr subnetworks (perhaps it is unrelated, though).

My `inn list --tree` looks like this:
```
interface: jenya (...)
  listening_port: 42500
  ip: 10.42.1.2
  10.42.0.0/16 jenya
    10.42.0.1/32 innernet-server
    | 10.42.0.1 innernet-server
    10.42.1.0/24 devices
    | 10.42.1.2 albus
    | 10.42.1.1 gellert
    | 10.42.1.3 pi
    10.42.2.0/24 others
    | 10.42.2.2 dobby
    | 10.42.2.1 lupin
    | 10.42.2.3 ron
    | 10.42.2.4 severus
```

The `gellert` and `pi` are located behind NAT1, the other hosts are located behind NAT2.

**The issue is**: the `inn fetch` throws an error:
```
inn fetch jenya
[*] fetching state from server.

[ERROR] http://10.42.0.1:51820/v1/user/state: Network Error: Network Error: Error encountered in the status line: timed out reading response
```

This happens from both `gellert` and `pi` nodes. The `inn fetch` functions correctly from all other nodes.
From `gellert` and `pi` I can ping the server 10.42.0.1, with a response of ~35ms. From `albus` I can ping the 10.42.0.1 with a response of ~9ms.

From `gellert` and `pi` I cannot ping anything else except the server. From `albus` I can reach devices all devices except `gellert` and `pi`.

Firewalls are temporary disabled on `gellert`, `pi` and the innernet-server.

**Comment by esovetkin** [comment]
Okay from a clean install I can see that for some I reason I still cannot access my innernet-server (from `gellert`). I guess something filtering out this traffic on my NAT1 side.

**Comment by mcginty** [comment]
Interesting.

1) What version of innernet are you using?
2) You say pings work - have you tried running another HTTP server on the innernet-server host a la `python -m http.server -b 10.42.0.1`, and then acecssing http://10.42.0.1:8000 from another peer? I'd just want to rule out something weird happening with the innernet-server binary itself.

**Comment by esovetkin** [comment]
Hi!
>     1. What version of innernet are you using?

on all nodes the version number is 1.2.0

>     2. You say pings work - have you tried running another HTTP server on the innernet-server host a la `python -m http.server -b 10.42.0.1`, and then acecssing http://10.42.0.1:8000 from another peer? I'd just want to rule out something weird happening with the innernet-server binary itself.

yes. this works as well!

**Comment by mcginty** [comment]
Interesting - there's been a large rewrite of the server (using `hyper` directly instead of `warp`) for v1.3.0 - would it be possible for you to try the latest beta and let me know if this is still happening?

**Comment by esovetkin** [comment]
> Thanks for the info. There's been a large rewrite of the server (using `hyper` directly instead of `warp`) for v1.3.0 - would it be possible for you to try the latest beta and let me know if this is still happening?

The issue persists with innernet-server with 1.3.0-beta.7. On the client side the innernet compiled with:
```
diff --git a/client/src/util.rs b/client/src/util.rs
index 55a421b..856377e 100644
--- a/client/src/util.rs
+++ b/client/src/util.rs
@@ -102,7 +102,7 @@ pub struct Api<'a> {
 impl<'a> Api<'a> {
     pub fn new(server: &'a ServerInfo) -> Self {
         let agent = AgentBuilder::new()
-            .timeout(Duration::from_secs(5))
+            .timeout(Duration::from_secs(50))
             .redirects(0)
             .build();
         Self { agent, server }
```
timeouts after 50 seconds.

> 
> Also, if there's anything interesting in the server logs `journalctl -u innernet-server@jenya` please paste them here to help debugging too. I've never had this specific issue before, personally.

There is nothing interesting in the logs. With version 1.3.0 the logs are:
```
May 21 04:06:47 ubuntu-s-1vcpu-1gb-fra1-01 systemd[1]: Started innernet server for jenya.
May 21 04:06:47 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[628]:  INFO  innernet_server::db > migrated db version from 0 to 1
May 21 04:06:47 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[628]:  INFO  innernet_server     > bringing up interface.
May 21 04:06:47 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[628]:  INFO  innernet_server     > 11 peers added to wireguard interface.
May 21 04:06:47 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[628]:  INFO  innernet_server     > innernet-server 1.3.0-beta.7 starting
```

with 1.2.0 version the logs are:
```
May 20 07:44:03 ubuntu-s-1vcpu-1gb-fra1-01 systemd[1]: Started innernet server for jenya.
May 20 07:44:03 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  innernet_server > bringing up interface.
May 20 07:44:03 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  innernet_server::endpoints > spawning endpoint watch thread.
May 20 07:44:03 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  innernet_server            > 6 peers added to wireguard interface.
May 20 07:44:03 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  innernet_server            > innernet-server 1.2.0 starting.
May 20 07:44:09 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  warp                       > 10.42.1.3:36552 "GET /v1/user/state HTTP/1.1" 200 "-" "ureq/2.1.0" 1.756084ms
May 20 07:44:09 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  warp                       > 10.42.1.1:49632 "GET /v1/user/state HTTP/1.1" 200 "-" "ureq/2.1.0" 713.607µs
May 20 07:44:14 ubuntu-s-1vcpu-1gb-fra1-01 innernet-server[624]:  INFO  warp                       > 10.42.1.3:36554 "GET /v1/user/state HTTP/1.1" 200 "-" "ureq/2.1.0" 548.81µs
...
```

I redid the network from scratch. From `gellert` in NAT1 (`gellert`, `pi`) I can see:
```
interface: jenya (...)    
  listening_port: 57032
  ip: 10.42.128.1
  10.42.0.0/16 jenya
    10.42.0.1/32 innernet-server
    | 10.42.0.1 innernet-server
    10.42.128.0/20 devices
    | 10.42.128.2 albus
    | 10.42.128.1 gellert
    | 10.42.128.3 lupin
```
The fetch stopped working on `gellert` while I was setting up the network. From `albus` in NAT2 I can see
```
interface: jenya (...)
  listening_port: 37604
  ip: 10.42.128.2
  10.42.0.0/16 jenya
    10.42.0.1/32 innernet-server
    | 10.42.0.1 innernet-server
    10.42.128.0/20 devices
    | 10.42.128.2 albus
    | 10.42.128.1 gellert
    | 10.42.128.4 dobby
    | 10.42.128.3 lupin
    | 10.42.128.5 severus
    | 10.42.128.7 ron
    | 10.42.128.8 pi
```
Interestingly, from `gellert` I can ping **only** `lupin`, `albus` and `innernet-server`. The `pi` failed fetch anything, but was registered in the innernet-server. I can see no hosts available from `pi`.

I don't know if this is important, but the ISP for NAT1 (`gellert`, `pi`) has a dual stack, with the same ipv4 address being given to many routers.

**Comment by mcginty** [comment]
On the innernet-server side, maybe record some outputs of `sudo wg show jenya` and track the external endpoints of the different peers as they send requests in - the shared IPv4 NAT1 setup sounds suspicious to me.

If you run another HTTP server, can all peers consistently reach it over time (all of them looping sending requests at the same time)? To be honest, it sounds like something in the NAT setup is causing peer endpoints to change so the server doesn't know what public endpoint to send responses back to.

**Comment by esovetkin** [comment]
> On the innernet-server side, maybe record some outputs of `sudo wg show jenya` and track the external endpoints of the different peers as they send requests in - the shared IPv4 NAT1 setup sounds suspicious to me.

Can I force the endpoints of some nodes to have an ipv6 address? I tried to set on `gellert` explicitly its ipv6 address via `inn override-endpoint jenya`, which has no effect on the server entry. After that I restarted the innernet on `gellert` and lost any connectivity to the `innernet-server`, with logs on `gellert` reporting: "[ERROR] http://10.42.0.1:51820/v1/user/state: Network Error: Network Error: Error encountered in the status line: timed out reading response"

**Comment by esovetkin** [comment]
I also noticed that, that `pi` cannot fetch anything, so the `inn list --tree` on `pi` is empty. But from `innernet-server` I can ping `pi` (10.42.128.8). At the same time I cannot ping `gellert` (10.42.128.1) from server, though `pi` and `gellert` are located on the same NAT.

**Comment by mcginty** [comment]
@esovetkin sorry for the delay in responding. Yeah, it makes sense that two machines within the same NAT have a hard time communicating with each other if the NAT doesn't provide NAT loopback functionality. Unfortunately, innernet doesn't have a *great* way to deal with this situation right now, besides you manually setting the endpoint to an internal IP address on the interfaces that are on the same NAT with a command like:

```sh
sudo wg set jenya peer <pubkey> endpoint 192.168.x.x:51820
```

**Comment by esovetkin** [comment]
Hi @mcginty, thanks for the help!I actually get a couple of problems here and there. For example, I have two identical nodes in the same NAT, with one being able to connect to the third node, and the other cannot.

It looks very weird to me, and I cannot really debug

**Comment by mcginty** [comment]
@esovetkin Right now, you can think of innernet as doing a very simple STUN-esque NAT holepunching technique.

Tailscale's article on NAT traversal is the best I've read to date: https://tailscale.com/blog/how-nat-traversal-works/, and they do a bit more fancy stuff to bust through particularly troublesome NAT configurations. I'm quite interested in implementing at least some of the things that Tailscale does as well in innernet, especially implementing something similar to ICE to attempt to connect to private addresses behind NATs, which it sounds like could help alleviate your connectivity issues.

**Comment by esovetkin** [comment]
thanks, @mcginty. that is indeed an excellent article!

**Comment by esovetkin** [comment]
hmm, I managed to resolve my issue by setting manually ipv6 address for the server.

To recall, most of my peers can operate only within ipv4 network. A single node (`gellert`) and a innernet-server have ipv6 addresses. `gellert` connected to a router, that shares a single ipv4 address among many peers. I generated gellert.toml with server ipv4 address, and during installation `fetch` fails.

However, if I manually change the innernet-server address to ipv6 everything seems to work.

`
inn install server.toml
wg set jenya  peer <peer_public_key> endpoint [ipv6_innernet_server]:51821
inn fetch jenya
  works as intended
`
